kaldb/src/main/java/com/slack/kaldb/server/RecoveryTaskFactory.java

```diff
@@ -0,0 +1,260 @@
+    List<SnapshotMetadata> deletedSnapshots = deleteStaleLiveSnapshots(snapshotsForPartition);
+
+    List<SnapshotMetadata> nonLiveSnapshotsForPartition =
+        snapshotsForPartition
+            .stream()
+            .filter(s -> !deletedSnapshots.contains(s))
+            .collect(Collectors.toUnmodifiableList());
+
+    // Get the highest offset that is indexed in durable store.
+    List<RecoveryTaskMetadata> recoveryTasks = recoveryTaskMetadataStore.listSync();
+    long highestDurableOffsetForPartition =
+        getHighestDurableOffsetForPartition(nonLiveSnapshotsForPartition, recoveryTasks);
+    LOG.info(
+        "The highest durable offset for partition {} is {}",
+        partitionId,
+        highestDurableOffsetForPartition);
+
+    if (highestDurableOffsetForPartition <= 0) {
+      LOG.info("There is no prior offset for this partition {}.", partitionId);
+      return highestDurableOffsetForPartition;
+    }
+
+    // The current head offset shouldn't be lower than the highest durable offset. If it is it
+    // means that we indexed more data than the current head offset. This is either a bug in the
+    // offset handling mechanism or the kafka partition has rolled over. We throw an exception
+    // for now, so we can investigate.
+    if (currentHeadOffsetForPartition < highestDurableOffsetForPartition) {
+      final String message =
+          String.format(
+              "The current head for the partition %d can't "
+                  + "be lower than the highest durable offset for that partition %d",
+              currentHeadOffsetForPartition, highestDurableOffsetForPartition);
+      LOG.error(message);
+      throw new IllegalStateException(message);
+    }
+
+    // The head offset for Kafka partition is the offset of the next message to be indexed. We
+    // assume that offset is passed into this function. The highest durable offset is the partition
+    // offset of the message that is indexed. Hence, the offset is incremented by 1 to get the
+    // next message.
+    long nextOffsetForPartition = highestDurableOffsetForPartition + 1;
+
+    // Create a recovery task if needed.
+    if (currentHeadOffsetForPartition - highestDurableOffsetForPartition > maxOffsetDelay) {
+      final long creationTimeEpochMs = Instant.now().toEpochMilli();
+      final String recoveryTaskName = "recoveryTask_" + partitionId + "_" + creationTimeEpochMs;
+      LOG.info(
+          "Recovery task needed. The current position {} and head location {} are higher than max"
+              + " offset {}",
+          highestDurableOffsetForPartition,
+          currentHeadOffsetForPartition,
+          maxOffsetDelay);
+      recoveryTaskMetadataStore.createSync(
+          new RecoveryTaskMetadata(
+              recoveryTaskName,
+              partitionId,
+              nextOffsetForPartition,
+              currentHeadOffsetForPartition - 1,
+              creationTimeEpochMs));
+      LOG.info(
```

Comment:
We probably need some additional metrics across this method - I think knowing the amount of recovery tasks created and the offsets/size (as tags?) would be helpful. That would allow us to see how often we're creating recovery tasks, and how big they are, etc.

Reply:
Done.